csharp/ql/src/semmle/code/csharp/exprs/Expr.qll

```diff
@@ -283,19 +282,148 @@ class IsTypeExpr extends IsExpr {
- */
-class IsPatternExpr extends IsTypeExpr {
+/** An `is` pattern expression, for example `x is string s`. */
+class IsPatternExpr extends IsExpr {
   LocalVariableDeclExpr typeDecl;
 
-  IsPatternExpr() { typeDecl = this.getChild(2) }
+  IsPatternExpr() { typeDecl = this.getChild(2) and not this instanceof IsRecursivePatternExpr }
 
   /**
    * Gets the local variable declaration in this `is` pattern expression.
    * For example `string s` in `x is string s`.
    */
   LocalVariableDeclExpr getVariableDeclExpr() { result = typeDecl }
+
+  /**
+   * Gets the type being accessed in this `is` expression, for example `string`
+   * in `x is string`.
+   */
+  Type getCheckedType() { result = getTypeAccess().getTarget() }
+
+  /**
+   * Gets the type access in this `is` expression, for example `string` in
+   * `x is string`.
+   */
+  TypeAccess getTypeAccess() { result = this.getChild(1) }
+}
+
+/** An `is` recursive pattern expression, for example `x is string { Length: 5 } s`. */
+class IsRecursivePatternExpr extends IsExpr {
+  RecursivePatternExpr recursivePattern;
+
+  IsRecursivePatternExpr() { recursivePattern = this.getChild(4) }
+
+  override string toString() { result = "... is { ... }" }
+
+  /**
+   * Gets the local variable declaration in this `is` pattern expression, if any.
+   * For example `s` in `x is string { Length: 0 } s`.
+   */
+  LocalVariableDeclExpr getVariableDeclExpr() { result = this.getChild(2) }
+
+  /**
+   * Gets the recursive pattern of this `is` pattern expression. For example,
+   * `{ Length: 5 }` in `x is string { Length: 5 } s`.
+   */
+  RecursivePatternExpr getRecursivePattern() { result = recursivePattern }
+}
+
+/**
+ * A recursive pattern expression, for example `string { Length: 5 } s`
+ */
+class RecursivePatternExpr extends Expr, @recursive_pattern_expr {
+  override string toString() { result = "{ ... }" }
+
+  /**
+   * Gets the position patterns of this recursive pattern, if any.
+   * For example, `(1,_)`.
+   */
+  PositionalPatternExpr getPositionalPatterns() { result = this.getChild(2) }
+
+  /**
+   * Gets the property patterns of this recursive pattern, if any.
+   * For example,
+   * `{ Length: 5 }` in `o is string { Length: 5 } s`
+   */
+  PropertyPatternExpr getPropertyPatterns() { result = this.getChild(3) }
+
+  /**
+   * Gets the type access of this recursive pattern, if any.
+   * For example, `string` in `string { Length: 5 }`
+   */
+  TypeAccess getTypeAccess() { result = this.getChild(1) }
+
+  /**
+   * Gets the local variable declaration of this recursive pattern, if any.
+   * For example, `s` in `string { Length: 5 } s`.
+   */
+  LocalVariableDeclExpr getVariableDeclExpr() { result = this.getChild(0) }
+}
+
+/** A property pattern. For example, `{ Length: 5 }`. */
+class PropertyPatternExpr extends Expr, @property_pattern_expr {
+  override string toString() { result = "{ ... }" }
+
+  /** Gets the `n`th pattern. */
+  Expr getPattern(int n) { result = this.getChild(n) }
```

Comment:
I wonder if we should have a new DB union type and `Expr` sub class for the return type of this? I assume it would be something like `@recursive_pattern_expr | @local_var_decl_expr | @discard_expr` ?

Reply:
It could also be a general constant, like `5` in `{ Length: 5 }`, or `false` in `(true, false)`.